Please push some empty commit to restart the checks, the PR is not mergeable because of failed special build check, though the check failure looks unrelated:
```
src/CMakeFiles/dbms.dir/Functions/FunctionsConversion.cpp.o -MF src/CMakeFiles/dbms.dir/Functions/FunctionsConversion.cpp.o.d -o src/CMakeFiles/dbms.dir/Functions/FunctionsConversion.cpp.o -c /build/src/Functions/FunctionsConversion.cpp
May 20 21:22:13 sccache: warning: The server looks like it shut down unexpectedly, compiling locally instead
May 20 21:22:13 sccache: Compile terminated by signal 9
May 20 21:22:13 ninja: build stopped: subcommand failed.
```

--- a/docs/en/operations/storing-data.md
+++ b/docs/en/operations/storing-data.md
@@ -421,6 +421,7 @@ Other parameters:
 * `skip_access_check` - If true, disk access checks will not be performed on disk start-up. Default value is `false`.
 * `read_resource` — Resource name to be used for [scheduling](/docs/en/operations/workload-scheduling.md) of read requests to this disk. Default value is empty string (IO scheduling is not enabled for this disk).
 * `write_resource` — Resource name to be used for [scheduling](/docs/en/operations/workload-scheduling.md) of write requests to this disk. Default value is empty string (IO scheduling is not enabled for this disk).
+* `metadata_keep_free_space_bytes` - the amount of free metadata disk space to be reserved.
 
 Examples of working configurations can be found in integration tests directory (see e.g. [test_merge_tree_azure_blob_storage](https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/test_merge_tree_azure_blob_storage/configs/config.d/storage_conf.xml) or [test_azure_blob_storage_zero_copy_replication](https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/test_azure_blob_storage_zero_copy_replication/configs/config.d/storage_conf.xml)).
 

--- a/src/Disks/ObjectStorages/MetadataStorageFactory.cpp
+++ b/src/Disks/ObjectStorages/MetadataStorageFactory.cpp
@@ -99,8 +99,10 @@ void registerMetadataStorageFromDisk(MetadataStorageFactory & factory)
     {
         auto metadata_path = config.getString(config_prefix + ".metadata_path",
                                               fs::path(Context::getGlobalContextInstance()->getPath()) / "disks" / name / "");
+        auto metadata_keep_free_space_bytes = config.getUInt64(config_prefix + ".metadata_keep_free_space_bytes", 0);
+
         fs::create_directories(metadata_path);
-        auto metadata_disk = std::make_shared<DiskLocal>(name + "-metadata", metadata_path, 0, config, config_prefix);
+        auto metadata_disk = std::make_shared<DiskLocal>(name + "-metadata", metadata_path, metadata_keep_free_space_bytes, config, config_prefix);
         auto key_compatibility_prefix = getObjectKeyCompatiblePrefix(*object_storage, config, config_prefix);
         return std::make_shared<MetadataStorageFromDisk>(metadata_disk, key_compatibility_prefix);
     });

--- a/tests/queries/0_stateless/02963_test_flexible_disk_configuration.sql
+++ b/tests/queries/0_stateless/02963_test_flexible_disk_configuration.sql
@@ -32,13 +32,35 @@ create table test (a Int32) engine = MergeTree() order by tuple()
 settings disk=disk(name='test3',
                    type = object_storage,
                    object_storage_type = s3,
+                   metadata_storage_type = local,
+                   metadata_keep_free_space_bytes = 1024,
+                   endpoint = 'http://localhost:11111/test/common/',
+                   access_key_id = clickhouse,
+                   secret_access_key = clickhouse);
+drop table test;
+
+create table test (a Int32) engine = MergeTree() order by tuple()
+settings disk=disk(name='test4',
+                   type = object_storage,
+                   object_storage_type = s3,
+                   metadata_storage_type = local,
+                   metadata_keep_free_space_bytes = 0,
+                   endpoint = 'http://localhost:11111/test/common/',
+                   access_key_id = clickhouse,
+                   secret_access_key = clickhouse);
+drop table test;
+
+create table test (a Int32) engine = MergeTree() order by tuple()
+settings disk=disk(name='test5',
+                   type = object_storage,
+                   object_storage_type = s3,
                    metadata_type = lll,
                    endpoint = 'http://localhost:11111/test/common/',
                    access_key_id = clickhouse,
                    secret_access_key = clickhouse); -- { serverError UNKNOWN_ELEMENT_IN_CONFIG }
 
 create table test (a Int32) engine = MergeTree() order by tuple()
-settings disk=disk(name='test4',
+settings disk=disk(name='test6',
                    type = object_storage,
                    object_storage_type = kkk,
                    metadata_type = local,
@@ -47,7 +69,7 @@ settings disk=disk(name='test4',
                    secret_access_key = clickhouse); -- { serverError UNKNOWN_ELEMENT_IN_CONFIG }
 
 create table test (a Int32) engine = MergeTree() order by tuple()
-settings disk=disk(name='test5',
+settings disk=disk(name='test7',
                    type = kkk,
                    object_storage_type = s3,
                    metadata_type = local,

